### ISSUE TITLE
RandomForest with configuration .all hyperparameters

### ISSUE THREAD

**Issue by nikkisingh111333** [oracle/tribuo#75]
Why random forest is not accepting my parameter n_estimator .i cannot find any documentation about random forest since its my one of the favourite algo and far good than cart.and one more thing ..how can i do cross validation and hyperparameter tuning ?
heres my config.json.
![Screenshot (38)](https://user-images.githubusercontent.com/49682940/97103769-a14d2f00-16d4-11eb-80af-f6c4130cec85.png)

**Comment by Craigacp** [comment]
To find the names of the configurable fields you can either inspect the source code, or use the `DescribeConfigurable` tool from OLCUT. For example:
```
java -cp tribuo-classification-experiments-4.0.1-jar-with-dependencies.jar com.oracle.labs.mlrg.olcut.config.DescribeConfigurable -n org.tribuo.common.tree.RandomForestTrainer -o true
```
Produces the output:
```
Class: org.tribuo.common.tree.RandomForestTrainer

Field Name   Type                                 Mandatory Redact Default Description
combiner     org.tribuo.ensemble.EnsembleCombiner true      false          The combination function to aggregate each ensemble member's outputs.
innerTrainer org.tribuo.Trainer                   true      false          The trainer to use for each ensemble member.
numMembers   int                                  true      false          The number of ensemble members to train.
seed         long                                 true      false          The seed for the RNG.

Example :
<?xml version="1.0" encoding="utf-8"?>
<!--OLCUT configuration file-->
<config>
<component name="example" type="org.tribuo.common.tree.RandomForestTrainer" export="false" import="false">
	<property name="seed" value="0"/>
	<property name="combiner" value="EnsembleCombiner-instance"/>
	<property name="innerTrainer" value="Trainer-instance"/>
	<property name="numMembers" value="0"/>
</component>
</config>
```

You can see the configurable variables and an example configuration snippet in xml, edn or json depending on the program arguments. We'll update the docs to reference this tool more clearly, but as Tribuo is open source you can see all this information in the code.

For the RandomForestTrainer the number of ensemble members is controlled by the `numMembers` field.

To perform hyperparameter tuning you'll need to measure the performance using either `CrossValidation` or a held out validation set, and then manually tune the parameters. It should be straightforward to write a grid search or random search for any specific trainer as you'll know the parameters you want to tune. We're considering how hyperparameter tuning would fit into Tribuo in a more general way, but we don't have anything to share at the moment.

**Comment by nikkisingh111333** [comment]
okk i understand numMembers is the num of trees and other params are same
as decision tree.
and how can i use crossvalidation is there any tribou api for cross
validation or i have to manually use For loop for training and testing my
model.?

**Comment by Craigacp** [comment]
You can use the [CrossValidation](https://tribuo.org/learn/4.0/javadoc/org/tribuo/evaluation/CrossValidation.html), and then aggregate the evaluations using [EvaluationAggregator](https://tribuo.org/learn/4.0/javadoc/org/tribuo/evaluation/EvaluationAggregator.html) to pull out the mean of whatever statistic you are maximising or minimising.

You will need to do the parameter selection yourself inside some kind of loop.

**Comment by Craigacp** [comment]
I've added some examples of the configuration description to the doc-fixes branch in the docs and in the configuration tutorial.

**Comment by nikkisingh111333** [comment]
thankyou adam...sorry for another question but i m still struggling with
random-forest algo.not accepting my hyperparameters in
config.json.impurity,maxdepth not accepting. dont know why?

**Comment by Craigacp** [comment]
Those are parameters of the tree algorithm you supply to the `RandomForestTrainer`. You need to pass it a reference to a tree algorithm elsewhere in the config, and an ensemble combiner (e.g. for classification use `VotingCombiner`). There is a complete example random forest classification configuration file [here](https://github.com/oracle/tribuo/blob/main/Classification/Experiments/configs/random-forest-example.xml).

**Comment by nikkisingh111333** [comment]
thanks a lot it helped me a lot.
